src/Stratis.Bitcoin.Tests/Base/ConsensusManagerBehaviorTests.cs

```diff
@@ -0,0 +1,278 @@
+        private int headersPresentedCalledTimes;
+
+        public ConsensusManagerBehaviorTests()
+        {
+            this.loggerFactory = new ExtendedLoggerFactory();
+            this.loggerFactory.AddConsoleWithFilters();
+
+            this.headers = ChainedHeadersHelper.CreateConsecutiveHeaders(100);
+        }
+
+        public class Counter
+        {
+            public int Value = 0;
+        }
+
+        private Mock<INetworkPeer> CreatePeerMock()
+        {
+            var peer = new Mock<INetworkPeer>();
+
+            var connection = new NetworkPeerConnection(Network.StratisMain, peer.Object, new TcpClient(), 0, (message, token) => Task.CompletedTask,
+                new DateTimeProvider(), this.loggerFactory, new PayloadProvider());
+
+            peer.SetupGet(networkPeer => networkPeer.Connection).Returns(connection);
+
+            var connectionParameters = new NetworkPeerConnectionParameters();
+            VersionPayload version = connectionParameters.CreateVersion(new IPEndPoint(1, 1), Network.StratisMain, new DateTimeProvider().GetTimeOffset());
+            version.Services = NetworkPeerServices.Network;
+
+            peer.SetupGet(x => x.PeerVersion).Returns(version);
+            peer.SetupGet(x => x.State).Returns(NetworkPeerState.HandShaked);
+            peer.SetupGet(x => x.MessageReceived).Returns(new AsyncExecutionEvent<INetworkPeer, IncomingMessage>());
+
+            this.stateChanged = new AsyncExecutionEvent<INetworkPeer, NetworkPeerState>();
+            this.messageReceived = new AsyncExecutionEvent<INetworkPeer, IncomingMessage>();
+
+            peer.Setup(x => x.StateChanged).Returns(() => this.stateChanged);
+            peer.Setup(x => x.MessageReceived).Returns(() => this.messageReceived);
+
+            return peer;
+        }
+
+        private ConsensusManagerBehavior CreateAndAttachBehavior(ChainedHeader consensusTip, out Counter getHeadersPayloadSentTimes,
+            List<BlockHeader> cache = null, ChainedHeader expectedPeerTip = null, NetworkPeerState peerState = NetworkPeerState.HandShaked,
+            Func<List<BlockHeader>, bool, ConnectNewHeadersResult> connectNewHeadersMethod = null)
+        {
+            // Chain
+            var chain = new ConcurrentChain(Network.StratisMain);
+            chain.SetTip(consensusTip);
+
+            // Ibd
+            var ibdState = new Mock<IInitialBlockDownloadState>();
+            ibdState.Setup(x => x.IsInitialBlockDownload()).Returns(() => this.IsIBD);
+
+            // Consensus manager
+            var cmMock = new Mock<IConsensusManager>();
+
+            cmMock.Setup(x => x.HeadersPresented(It.IsAny<INetworkPeer>(), It.IsAny<List<BlockHeader>>(), It.IsAny<bool>()))
+                .Returns((INetworkPeer p, List<BlockHeader> presentedHeaders, bool triggerDownload) =>
+            {
+                this.headersPresentedCalledTimes++;
```

Comment:
Is here not another problem with parallel execution?

Reply:
For some reason it never crushed the tests. So didn't change it.